netam/molevol.py

```diff
@@ -512,52 +536,181 @@ def log_pcp_probability(log_branch_length: torch.Tensor):
     return log_pcp_probability
 
 
-def optimize_branch_length(
-    log_prob_fn,
-    starting_branch_length,
-    learning_rate=0.1,
-    max_optimization_steps=1000,
-    optimization_tol=1e-3,
-    log_branch_length_lower_threshold=-10.0,
+def find_bracket(
+    func,
+    start,
+    suggested_middle,
+    end,
+    epsilon=1e-10,
+    brute_linear_step=0.001,
 ):
-    log_branch_length = torch.tensor(np.log(starting_branch_length), requires_grad=True)
-
-    optimizer = optim.Adam([log_branch_length], lr=learning_rate)
-    prev_log_branch_length = log_branch_length.clone()
-
-    step_idx = 0
+    """Find bracket for optimization. Requirements are start < middle < end and
+    func(start) > func(middle) < func(end)
 
-    for step_idx in range(max_optimization_steps):
-        # For some PCPs, the optimizer works very hard optimizing very tiny branch lengths.
-        if log_branch_length < log_branch_length_lower_threshold:
-            break
+    Attempts to use scipy.optimize.bracket, then falls back to brute force search if necessary.
+    Args:
+        func: The function to optimize.
+        start: The minimum value expected for the minimizer.
+        suggested_middle: A suggested starting point for the middle value.
+        end: The maximum value expected for the minimizer.
+        epsilon: A small value to avoid false positives finding minimum function values in brute force search.
+        brute_linear_step: Step size for brute force search.
 
-        optimizer.zero_grad()
+    Returns:
+        A tuple containing the start, middle, and end points of the bracket.
+        start and end may be outside the range of the passed start and end, but middle point will be contained in the passed range.
+    """
+    if suggested_middle <= start:
+        suggested_middle = np.log(np.exp(start) + 2 * linear_step)
+    if suggested_middle >= end:
+        end = suggested_middle + 2
+
+    try_vals = [
+        [suggested_middle - 0.1, suggested_middle + 0.1],
+        [
+            np.log(max(epsilon, np.exp(suggested_middle) - 0.01)),
+            np.log(np.exp(suggested_middle) + 0.01),
+        ],
+        [suggested_middle - 0.5, suggested_middle + 0.5],
+        [suggested_middle - 1.0, suggested_middle],
+        [suggested_middle + 1.0, suggested_middle],
+        [suggested_middle - 2.0, suggested_middle],
+        [suggested_middle + 2.0, suggested_middle],
+        [(end - start) / 2, suggested_middle],
```

Comment:
This is absurd, and in my tests the first option works for all 50k pcps I tested. However, the first option was the result of much trial and error including the other options on this list, so I leave them all to hopefully handle future insane, unforeseen  PCPs and suggested optima.